### PR TITLE
Completed pattern matching code listings

### PR DESCRIPTION
**Pull request by MarkMichaelis**

### CODE REVIEW

File: src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter07.Listing07_26.Tests;
+
+[TestClass]
+public class ProgramTests
+{
+    string DataFile = "data.dat";
+    const string EncryptedFileName = "temp.out";
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs]:
```suggestion
    private string DataFile = "data.dat";
    private const string EncryptedFileName = "temp.out";
```

File: src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+    [TestMethod]
+    public async Task Main_EncryptFileUsingIsOperator_Success()
+    {
+        await InvokeMainWithEncriptAction(
+            Program.Main);
+    }
+
+    private async Task InvokeMainWithEncriptAction(Action<string[]> action)
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs]: I don't mind these helper methods, but often I will push back on their existence when they are only invoked once. Does it provide value beyond just putting these in the test method itself.

File: src/Chapter07.Tests/Listing07.28.PositionalPatternMatching.Tests.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System.Drawing;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter07.Listing07_28.Tests;
+
+[TestClass]
+public class ProgramTests
+{
+    const bool IsVgaVisible = true;
+    const bool NotVgaVisible = false;
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.28.PositionalPatternMatching.Tests.cs:10]:
```suggestion
    private const bool IsVgaVisible = true;
    private const bool NotVgaVisible = false;
```

File: src/Chapter07.Tests/Listing07.29.PropertyPatternMatching.Tests.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter07.Listing07_29.Tests;
+
+[TestClass]
+public class ProgramTests
+{
+    const bool IsValid = true;
+    const bool NotValid = false;
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.29.PropertyPatternMatching.Tests.cs:9]:
```suggestion
    private const bool IsValid = true;
    private const bool NotValid = false;
```

File: src/Chapter07.Tests/Listing07.32.Tests.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    [TestMethod]
+    [DataRow("help")]
+    [DataRow("h")]
+    [DataRow("?")]
+    [DataRow("<invalid>")]
+    public void ParseHelpCommand(string command)
+    {
+        foreach (string item in PrefixCommands(command))
+        {
+            command = $"{item}";
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.32.Tests.cs]:
```suggestion
            command = item;
```

File: src/Chapter07.Tests/Listing07.32.Tests.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    [TestMethod]
+    [DataRow("cat", @"..\SampleFile.txt")]
+    [DataRow("c", @"..\SampleFile.txt")]
+    public void ParseCatCommand(string command, string fileName)
+    {
+        foreach (string item in PrefixCommands(command))
+        {
+            command = $"{item}";
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.32.Tests.cs:32]:
```suggestion
            command = item;
```

File: src/Chapter07.Tests/Listing07.32.Tests.cs [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+
+    [TestMethod]
+    [DataRow("copy", @"..\source.txt", @"..\target.txt")]
+    public void ParseCopyCommand(string command, string sourceFile, string targetFile)
+    {
+        foreach (string item in PrefixCommands(command))
+        {
+            command = $"{item}";
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.32.Tests.cs:45]:
```suggestion
            command = item;
```

File: src/Chapter07.Tests/Listing07.32.Tests.cs [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+            {
+                commands.Add($"{item}{command}");
+            }
+        }
+        else
+        {
+            commands.Add($"--{command}");
+        }
+        return commands;
+    }
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.32.Tests.cs]:
```suggestion
    private static IEnumerable<string> PrefixCommands(string command)
    {
        if (command.Length == 1)
        {
            yield return $"/{command}";
            yield return $"-{command}";
        }
        else
        {
            yield return $"--{command}";
        }
    }
```

File: src/Chapter07/Listing07.11.OverrideVersusNewModifier.cs [hunk 9]
```diff
@@ -1,3 +1,4 @@
+#pragma warning disable IDE0059 // Unnecessary assignment of a value
```
Comment (Keboo) [inline src/Chapter07/Listing07.11.OverrideVersusNewModifier.cs:1]: It looks like there is another warning in this file, should we be suppressing it as well?

File: src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    public string Id { get; }
+    #endregion EXCLUDE
+}
+
+public class Program
+{
+    static object? GetObjectById(string id)
+    {
+        #region EXCLUDE
+        if(id is null) throw new ArgumentNullException("id");
```
Comment (Keboo) [inline src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs:38]:
```suggestion
        ArgumentNullException.ThrowIfNull(id);
```

File: src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+        if (id.StartsWith(nameof(Employee)))
+        {
+            return new Employee("Inigo", "Montoya", id.Replace(nameof(Employee), ""));
```
Comment (Keboo) [inline src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs]: I would generally not do a replace after a startswith call. I am sure it works in this simple case, but in real apps this opens up bugs when the string contains more than one instance of "Employee"

File: src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        #region HIGHLIGHT
+        else if (entity is Employee)
+        #endregion HIGHLIGHT
+        {
+            throw new InvalidOperationException(
+                $"Entity ({entity.GetType().FullName
+                    }) is unexpectedly an {nameof(Employee)} object.");
+        }
+        else if(entity is null)
```
Comment (Keboo) [inline src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs:86]:
```suggestion
        else if (entity is null)
```

File: src/Chapter07/Listing07.20.SwitchStatement.cs [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+        switch (input)
+        {
+            case DateTime datetime:
+                return TimeOnly.FromDateTime(datetime);
+            case DateTimeOffset datetimeOffset:
+                return TimeOnly.FromDateTime(datetimeOffset.DateTime);
+            case string dateText:
+                return TimeOnly.Parse(dateText);
+            case null:
+                throw new Exception();
```
Comment (Keboo) [inline src/Chapter07/Listing07.20.SwitchStatement.cs]:
```suggestion
                throw new ArgumentNullException(nameof(input));
```

File: src/Chapter07/Listing07.21.UsingSwtichExpressionToReturnValue.cs [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    public static TimeOnly GetTime(object input) =>
+        input switch
+        {
+            DateTime datetime
+                => TimeOnly.FromDateTime(datetime),
+            DateTimeOffset datetimeOffset
+                    => TimeOnly.FromDateTime(datetimeOffset.DateTime),
+            string dateText => TimeOnly.Parse(
+                dateText),
+            null => throw new Exception(),
```
Comment (Keboo) [inline src/Chapter07/Listing07.21.UsingSwtichExpressionToReturnValue.cs]:
```suggestion
            null => throw new ArgumentNullException(nameof(input)),
```

File: src/Chapter07/Listing07.22.ConstantPatternWithSwitchExpression.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+            }
+        }
+        return false;
+    }
+
+    public static bool TryGetSolstice(DateTime date,
+        [NotNullWhen(true)] out string? solstice
+    )
+    {
+        if ((decimal)date.Day is 21)
```
Comment (Keboo) [inline src/Chapter07/Listing07.22.ConstantPatternWithSwitchExpression.cs]:
```suggestion
        if (date.Day is 21)
```

File: src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+        const int command = 0;
+        const int fileName = 1;
+        const string dataFile = "data.dat";
+
+        // ...
+        #region HIGHLIGHT
+        if ((args.Length, args[command]) is (1, "show"))
+        #endregion HIGHLIGHT
+        {
+            Console.WriteLine(File.ReadAllText(dataFile));
```
Comment (Keboo) [inline src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs:22]: In a real app I would tell people to never call things like `ReadAllText` blinding on a user provided file.

What about changing this to just `Console.WriteLine($"File name: {dataFile}");`

File: src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        }
+        #region HIGHLIGHT
+        else if ((args.Length, args[command].ToLower())
+            is (2, "encrypt"))
+        #endregion HIGHLIGHT
+        {
+            string data = File.ReadAllText(dataFile);
+            File.WriteAllText(
+                args[fileName], Encrypt(data).ToString());
```
Comment (Keboo) [inline src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs:31]: Again, in a real app I would strong discourage blind reading of a user provided path. Perhaps something similar where we just echo out the information?

File: src/Chapter07/Listing07.27.TuplePatternMatchingWithSwitchStatement.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        const int action = 0;
+        const int fileName = 1;
+        const string dataFile = "data.dat";
+
+        // ...
+        #region HIGHLIGHT
+        switch ((args.Length, args[action]))
+        {
+            case (1, "show"):
+                Console.WriteLine(File.ReadAllText(dataFile));
```
Comment (Keboo) [inline src/Chapter07/Listing07.27.TuplePatternMatchingWithSwitchStatement.cs:17]: Same thing here.

File: src/Chapter07/Listing07.27.TuplePatternMatchingWithSwitchStatement.cs [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+                break;
+            case (2, "encrypt"):
+        #endregion HIGHLIGHT
+                {
+                    string data = File.ReadAllText(dataFile);
```
Comment (Keboo) [inline src/Chapter07/Listing07.27.TuplePatternMatchingWithSwitchStatement.cs:22]: Same thing here.

File: src/Chapter07/Listing07.29.PropertyPatternMatching.cs [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+        (Id, Name, Role) = (id, name, role);
+}
+
+
+public class ExpenseItem
+{
+    public int Id { get; set; }
+    public string ItemName { get; set; }
+    public decimal CostAmount { get; set; }
+    public DateTime ExpenseDate { get; set; } = DateTime.Now;
```
Comment (Keboo) [inline src/Chapter07/Listing07.29.PropertyPatternMatching.cs]: Default value appears to always get overridden in the constructor.
```suggestion
    public DateTime ExpenseDate { get; set; }
```

File: src/Chapter07/Listing07.32.ListPatternMatching.cs [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+            case []:
+                // No command line arguments
+
+            default:
+                DisplayHelp();
+                break;
+        }
+    }
+
+    static bool ExecuteCommand(string command, string[] args) =>
```
Comment (Keboo) [inline src/Chapter07/Listing07.32.ListPatternMatching.cs]:
```suggestion
    private static bool ExecuteCommand(string command, string[] args) =>
```

File: src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs (hunk 2 above)
Comment (MarkMichaelis) [reply src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs]: Agreed.... thanks. I split the code into two files and didn't complete the refactor.

File: src/Chapter07/Listing07.11.OverrideVersusNewModifier.cs (hunk 9 above)
Comment (MarkMichaelis) [reply src/Chapter07/Listing07.11.OverrideVersusNewModifier.cs:1]: No... this is part of the discussion in the text. Good question, however, it forced me to change my approach.

File: src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs (hunk 16 above)
Comment (MarkMichaelis) [reply src/Chapter07/Listing07.26.TuplePatternMatchingWithIsOperator.cs:22]: Hmm. You are right. But the method catalogs ("cat") the file. Without async, this seems like the simplest. We can discuss if you wish.

File: src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs [hunk 22]
```diff
@@ -0,0 +1,58 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace AddisonWesley.Michaelis.EssentialCSharp.Chapter07.Listing07_26.Tests;
+
+static class TuplePatternMatchingTests
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs]:
```suggestion
internal static class TuplePatternMatchingTests
```

File: src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+{
+    static string DataFile = "./data.dat";
+    const string EncryptedFileName = "./temp.out";
```
Comment (Keboo) [inline src/Chapter07.Tests/Listing07.26.TuplePatternMatchingWithIsOperator.Tests.cs]:
```suggestion
    private const string DataFile = "./data.dat";
    private const string EncryptedFileName = "./temp.out";
```

File: src/Chapter07.Tests/Listing07.32.Tests.cs (hunk 5 above)
Comment (Keboo) [reply src/Chapter07.Tests/Listing07.32.Tests.cs]: Intentionally resolved without change?

File: src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        Id = id ?? throw new ArgumentNullException(nameof(id));
+    }
+    public string Id { get; }
+    #endregion EXCLUDE
+}
+
+public class Program
+{
+    static object? GetObjectById(string id)
```
Comment (Keboo) [inline src/Chapter07/Listing07.19.TypeCheckingWithIsOperator.cs]:
```suggestion
    private static object? GetObjectById(string id)
```

**Review by Keboo — APPROVED** [review]

**Pull request merged**